modules/system/assets/js/snowboard/main/Snowcart.js

```diff
@@ -0,0 +1,347 @@
+
+            // Call event handler methods for all plugins, if they have a method specified for the event.
+            plugin.getInstances().forEach((instance) => {
+                // If a plugin has cancelled the event, no further plugins are considered.
+                if (cancelled) {
+                    return;
+                }
+
+                if (!instance[listenMethod]) {
+                    throw new Error(`Missing "${listenMethod}" method in "${name}" plugin`);
+                }
+
+                if (instance[listenMethod](...args) === false) {
+                    cancelled = true;
+                }
+            });
+        });
+
+        return !cancelled;
+    }
+
+    /**
+     * Calls a global event to all registered plugins, expecting a Promise to be returned by all.
+     *
+     * This collates all plugins responses into one large Promise that either expects all to be resolved, or one to reject.
+     * If no listeners are found, a resolved Promise is returned.
+     *
+     * @param {string} eventName
+     */
+    globalPromiseEvent(eventName) {
+        /* develblock:start */
+        this.debug(`Calling global promise event "${eventName}"`);
+        /* develblock:end */
+
+        // Find out which plugins listen to this event - if none listen to it, return a resolved promise.
+        const listeners = this.listensToEvent(eventName);
+        if (listeners.length === 0) {
+            return Promise.resolve();
+        }
+
+        const promises = [];
+        let args = Array.from(arguments);
+        args.shift();
+
+        listeners.forEach((name) => {
+            const plugin = this.getPlugin(name);
+
+            if (plugin.isFunction()) {
+                return;
+            }
+            if (plugin.isSingleton() && plugin.getInstances().length === 0) {
+                plugin.initialiseSingleton();
+            }
+
+            const listenMethod = plugin.callMethod('listens')[eventName];
+
+            // Call event handler methods for all plugins, if they have a method specified for the event.
+            plugin.getInstances().forEach((instance) => {
+                const instancePromise = instance[listenMethod](...args);
+                if (instancePromise instanceof Promise === false) {
+                    return;
+                }
+
+                promises.push(instancePromise);
+            });
+        });
+
+        if (promises.length === 0) {
+            return Promise.resolve();
+        }
+
+        return Promise.all(promises);
+    }
+
+    /* develblock:start */
+    /**
+     * Log a debug message.
+     *
+     * These messages are only shown when debugging is enabled.
+     *
+     * @returns {void}
+     */
+    debug() {
+        if (!this.debugEnabled) {
+            return;
+        }
+
+        console.log("%c[Snowboard]", "color: rgb(45, 167, 199);", ...arguments);
+    }
+    /* develblock:end */
```

Comment:
Debug function should be available in prod builds for plugins that don't have dev / prod builds but want to only log messages when app.debug is enabled.